tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+                .ok_or(TransactionError::AccountInterfaceNotFound(id)),
+            NoteTarget::Procedures(procs) => Ok(procs),
+        }
+    }
+
+    /// Compiles a transaction which executes the provided notes against the specified account.
+    ///
+    /// The account is assumed to have been previously loaded into this compiler.
+    pub fn compile_transaction(
+        &mut self,
+        account_id: AccountId,
+        notes: Vec<Note>,
+        tx_script: Option<ProgramAst>,
+    ) -> Result<CompiledTransaction, TransactionError> {
+        // Check if the account has been loaded into the [TransactionCompiler]
+        if !self.account_procedures.contains_key(&account_id) {
+            return Err(TransactionError::AccountInterfaceNotFound(account_id));
+        }
+
+        // Transaction must contain at least one input note
+        if notes.is_empty() {
+            return Err(TransactionError::NoNotesProvided);
+        }
+
+        // Create the [AssemblyContext] for compilation of the transaction program
+        let mut assembly_context = AssemblyContext::new(AssemblyContextType::Program);
+
+        // Create note tree
+        let note_program_tree =
+            self.create_note_program_tree(&account_id, &notes, &mut assembly_context)?;
+
+        // Create the transaction script [CodeBlock].  If no tx_script is provided we use a
+        // Noop span block.
+        let tx_script = match tx_script {
+            Some(tx_script) => self
+                .assembler
+                .compile_in_context(tx_script, &mut assembly_context)
+                .map_err(TransactionError::CompileTxScriptFailed)?,
+            None => CodeBlock::new_span(vec![Operation::Noop]),
+        };
+
+        // Merge transaction script code block and epilogue code block
+        let tx_script_and_epilogue =
+            CodeBlock::new_join([tx_script.clone(), self.epilogue.clone()]);
+
+        // Merge prologue and note script tree
+        let prologue_and_notes = CodeBlock::new_join([self.prologue.clone(), note_program_tree]);
+
+        // Merge prologue, note tree, tx script and epilogue
+        let program_root = CodeBlock::new_join([prologue_and_notes, tx_script_and_epilogue]);
+
+        // Create [CodeBlockTable] from [AssemblyContext]
+        let cb_table = self
+            .assembler
+            .build_cb_table(assembly_context)
+            .map_err(TransactionError::BuildCodeBlockTableFailed)?;
+
+        // Create transaction program
+        let program = Program::with_kernel(program_root, self.tx_kernel.clone(), cb_table);
+
+        // Create compiled transaction
+        // TODO: we always have a digest for a transaction script (noop code block root if no user
+        // tx scrip is provided) - should we we change the [CompiledTransaction] struct to take a
+        // digest instead of a Option<Digest> fro the tx script.
+        Ok(CompiledTransaction::new(account_id, notes, Some(tx_script.hash()), program))
```

Comment:
I'm of a split mind here: on the one hand, it does simplify the interface. On the other, I wonder if we want to have a simple way to determine whether a given transaction has a non-empty `tx_script`.

Reply:
An alternative would be to check if the `tx_script` hash is equivalent to a noop block. However, I think we should probably leave it as it is for now.